doc/user-guide/testing.rst

```diff
@@ -0,0 +1,178 @@
+examples.
+
+.. ipython:: python
+    :okexcept:
+
+    # generate a DataArray with shape (3, 4), but all other details still arbitrary
+    xrst.dataarrays(
+        data=xrst.np_arrays(shape=(3, 4), dtype=np.dtype("int32"))
+    ).example()
+
+This also works with strategies defined in other packages, for example the ``chunks`` strategy defined in
+``dask.array.strategies``.
+
+.. warning::
+    Passing multiple different strategies to the same constructor can lead to poor example generation performance.
+
+    This is because in order to construct a valid xarray object to return, our strategies must check that the
+    variables / dimensions / coordinates are mutually compatible. We do this using ``hypothesis.assume``, which throws
+    away any generated examples not meeting the required condition.
+
+    Therefore if you pass multiple custom strategies to a strategy constructor which are not compatible in enough cases,
+    most of the examples they generate will be mutually incompatible. This will likely lead to poor example generation
+    performance, manifesting as a ``hypothesis.errors.FailedHealthCheck`` being raised. For example:
+
+    .. code-block::
+
+        @given(st.data())
+        def test_something_else_inefficiently(data):
+            arrs = xrst.np_arrays()  # generates arrays of any shape
+            dims = xrst.dimension_names()  # generates lists of any number of dimensions
+
+            # Drawing examples from this strategy is likely to have poor performance
+            var = data.draw(xrst.variables(data=arrs, dims=dims))
+
+            assert ...
+
+    Here we have passed custom strategies which won't often be compatible: only rarely will the array's ``ndims``
+    correspond to the number of dimensions drawn.
+
+    To avoid this problem either allow xarray's strategies to automatically generate compatible data for you, or be more
+    selective about cases when passing multiple custom strategies to the same constructor.
+
+
+Fixing Arguments
+~~~~~~~~~~~~~~~~
+
+If you want to fix one aspect of the data structure, whilst allowing variation in the generated examples
+over all other aspects, then use ``hypothesis.strategies.just()``.
+
+.. ipython:: python
+
+    import hypothesis.strategies as st
+
+    # Generates only dataarrays with dimensions ["x", "y"]
+    xrst.dataarrays(dims=st.just(["x", "y"])).example()
+
+(This is technically another example of chaining strategies - ``hypothesis.strategies.just`` is simply a special
+strategy that just contains a single example.)
+
+To fix the length of dimensions you can instead pass `dims` as a mapping of dimension names to lengths
+(i.e. following xarray objects' ``.sizes()`` property), e.g.
+
+.. ipython:: python
+
+    # Generates only dataarrays with dimensions ["x", "y"], of lengths 2 & 3 respectively
+    xrst.dataarrays(dims=st.just({"x": 2, "y": 3})).example()
```

Comment:
Good point - I think that most users who want custom strategies will either want:

1) Some specific set of dimensions / lengths, in which case documenting this is very important as you say,
2) To change the type of the underlying array (e.g. to sparse instead of numpy). This is supposed to be handled by the `convert` argument.

Reply:
Added example.